// aws-cdk/lib/cli/io-host/cli-io-host.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CliIoHost = void 0;
const util = require("node:util");
const cloud_assembly_schema_1 = require("@aws-cdk/cloud-assembly-schema");
const toolkit_lib_1 = require("@aws-cdk/toolkit-lib");
const chalk = require("chalk");
const promptly = require("promptly");
const api_private_1 = require("../../../lib/api-private");
const deploy_1 = require("../../commands/deploy");
const collect_telemetry_1 = require("../telemetry/collect-telemetry");
const error_1 = require("../telemetry/error");
const messages_1 = require("../telemetry/messages");
const session_1 = require("../telemetry/session");
const endpoint_sink_1 = require("../telemetry/sink/endpoint-sink");
const file_sink_1 = require("../telemetry/sink/file-sink");
const funnel_1 = require("../telemetry/sink/funnel");
const ci_1 = require("../util/ci");
/**
 * A simple IO host for the CLI that writes messages to the console.
 */
class CliIoHost {
    /**
     * Returns the singleton instance
     */
    static instance(props = {}, forceNew = false) {
        if (forceNew || !CliIoHost._instance) {
            CliIoHost._instance = new CliIoHost(props);
        }
        return CliIoHost._instance;
    }
    /**
     * Returns the singleton instance if it exists
     */
    static get() {
        return CliIoHost._instance;
    }
    constructor(props = {}) {
        /**
         * Configure the target stream for notices
         *
         * (Not a setter because there's no need for additional logic when this value
         * is changed yet)
         */
        this.noticesDestination = 'stderr';
        this._progress = deploy_1.StackActivityProgress.BAR;
        // Corked Logging
        this.corkedCounter = 0;
        this.corkedLoggingBuffer = [];
        // Message listeners, keyed by message code. See `on`/`once`/`rewrite`/`rewriteOnce`.
        this.messageListeners = new Map();
        this.currentAction = props.currentAction ?? 'none';
        this.isTTY = props.isTTY ?? process.stdout.isTTY ?? false;
        this.logLevel = props.logLevel ?? 'info';
        this.isCI = props.isCI ?? (0, ci_1.isCI)();
        this.requireDeployApproval = props.requireDeployApproval ?? cloud_assembly_schema_1.RequireApproval.BROADENING;
        this.stackProgress = props.stackProgress ?? deploy_1.StackActivityProgress.BAR;
        this.autoRespond = props.autoRespond ?? false;
        // Stack-activity messages are handled by the activity printer rather than
        // written to a stream. This is wired up as message listeners.
        this.routeStackActivityToPrinter();
    }
    async startTelemetry(args, context, proxyAgent) {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const config = require('../cli-type-registry.json');
        const validCommands = Object.keys(config.commands);
        const cmd = args._[0];
        if (!validCommands.includes(cmd)) {
            // the user typed in an invalid command - no need for telemetry since the invocation is going to fail
            // imminently anyway.
            await this.asIoHelper().defaults.trace(`Session instantiated with an invalid command (${cmd}). Not starting telemetry.`);
            return;
        }
        let sinks = [];
        const telemetryFilePath = args['telemetry-file'];
        if (telemetryFilePath) {
            try {
                sinks.push(new file_sink_1.FileTelemetrySink({
                    ioHost: this,
                    logFilePath: telemetryFilePath,
                }));
                await this.asIoHelper().defaults.trace('File Telemetry connected');
            }
            catch (e) {
                await this.asIoHelper().defaults.trace(`File Telemetry instantiation failed: ${e.message}`);
            }
        }
        const telemetryEndpoint = process.env.TELEMETRY_ENDPOINT ?? 'https://cdk-cli-telemetry.us-east-1.api.aws/metrics';
        if ((0, collect_telemetry_1.canCollectTelemetry)(args, context) && telemetryEndpoint) {
            try {
                sinks.push(new endpoint_sink_1.EndpointTelemetrySink({
                    ioHost: this,
                    agent: proxyAgent,
                    endpoint: telemetryEndpoint,
                }));
                await this.asIoHelper().defaults.trace('Endpoint Telemetry connected');
            }
            catch (e) {
                await this.asIoHelper().defaults.trace(`Endpoint Telemetry instantiation failed: ${e.message}`);
            }
        }
        else {
            await this.asIoHelper().defaults.trace('Endpoint Telemetry NOT connected');
        }
        if (sinks.length > 0) {
            this.telemetry = new session_1.TelemetrySession({
                ioHost: this,
                client: new funnel_1.Funnel({ sinks }),
                arguments: args,
                context: context,
            });
        }
        await this.telemetry?.begin();
    }
    /**
     * Update the stackProgress preference.
     */
    set stackProgress(type) {
        this._progress = type;
    }
    /**
     * Gets the stackProgress value.
     *
     * This takes into account other state of the ioHost,
     * like if isTTY and isCI.
     */
    get stackProgress() {
        // We can always use EVENTS
        if (this._progress === deploy_1.StackActivityProgress.EVENTS) {
            return this._progress;
        }
        // if a debug message (and thus any more verbose messages) are relevant to the current log level, we have verbose logging
        const verboseLogging = (0, api_private_1.isMessageRelevantForLevel)({ level: 'debug' }, this.logLevel);
        if (verboseLogging) {
            return deploy_1.StackActivityProgress.EVENTS;
        }
        // On Windows we cannot use fancy output
        const isWindows = process.platform === 'win32';
        if (isWindows) {
            return deploy_1.StackActivityProgress.EVENTS;
        }
        // On some CI systems (such as CircleCI) output still reports as a TTY so we also
        // need an individual check for whether we're running on CI.
        // see: https://discuss.circleci.com/t/circleci-terminal-is-a-tty-but-term-is-not-set/9965
        const fancyOutputAvailable = this.isTTY && !this.isCI;
        if (!fancyOutputAvailable) {
            return deploy_1.StackActivityProgress.EVENTS;
        }
        // Use the user preference
        return this._progress;
    }
    get defaults() {
        return this.asIoHelper().defaults;
    }
    asIoHelper() {
        return (0, api_private_1.asIoHelper)(this, this.currentAction);
    }
    /**
     * Executes a block of code with corked logging. All log messages during execution
     * are buffered and only written when all nested cork blocks complete (when CORK_COUNTER reaches 0).
     * The corking is bound to the specific instance of the CliIoHost.
     *
     * @param block - Async function to execute with corked logging
     * @returns Promise that resolves with the block's return value
     */
    async withCorkedLogging(block) {
        this.corkedCounter++;
        try {
            return await block();
        }
        finally {
            this.corkedCounter--;
            if (this.corkedCounter === 0) {
                // Process each buffered message through notify
                for (const ioMessage of this.corkedLoggingBuffer) {
                    await this.notify(ioMessage);
                }
                // remove all buffered messages in-place
                this.corkedLoggingBuffer.splice(0);
            }
        }
    }
    /**
     * Register a listener that is invoked for every message with the given code.
     *
     * The listener may return a `MessageListenerResult` to update the message
     * text and/or prevent the default processing (writing it to a stream);
     * returning nothing leaves the message untouched. Returns a function that
     * removes the listener again.
     *
     * @example
     * const dispose = ioHost.on(IO.CDK_TOOLKIT_I2901, (msg) => {
     *   myCount += msg.data.stacks.length;
     * });
     */
    on(code, listener) {
        return this.addMessageListener(code.code, { once: false, fn: listener });
    }
    /**
     * Like `on`, but the listener is automatically removed after it has been
     * invoked once.
     */
    once(code, listener) {
        return this.addMessageListener(code.code, { once: true, fn: listener });
    }
    /**
     * Register a formatter that replaces the printed text of messages with the
     * given code. This lets a caller define _how_ a toolkit message is presented
     * without the IoHost needing to know about it.
     *
     * Syntactic sugar for an `on` listener that returns `{ message }`. Returns a
     * function that removes the formatter again.
     *
     * @example
     * const dispose = ioHost.rewrite(IO.CDK_TOOLKIT_I2901, (msg) =>
     *   serializeStructure(msg.data.stacks, true));
     */
    rewrite(code, formatter) {
        return this.on(code, (msg) => ({ message: formatter(msg) }));
    }
    /**
     * Like `rewrite`, but the formatter is automatically removed after it has
     * been applied once.
     */
    rewriteOnce(code, formatter) {
        return this.once(code, (msg) => ({ message: formatter(msg) }));
    }
    /**
     * Add a listener to the registry and return a function that removes it.
     */
    addMessageListener(code, listener) {
        const listeners = this.messageListeners.get(code) ?? [];
        listeners.push(listener);
        this.messageListeners.set(code, listeners);
        return () => {
            const index = listeners.indexOf(listener);
            if (index >= 0) {
                listeners.splice(index, 1);
            }
        };
    }
    /**
     * Run all registered listeners for a message's code, in registration order.
     *
     * A listener may update the message text (which is passed on to subsequent
     * listeners and the rest of the pipeline) and/or request that the default
     * processing be skipped. `once` listeners are removed after they have run.
     *
     * Returns the (possibly text-updated) message and whether any listener
     * prevented the default processing.
     */
    applyMessageListeners(msg) {
        let current = msg;
        let preventDefault = false;
        const listeners = msg.code ? this.messageListeners.get(msg.code) : undefined;
        if (listeners && listeners.length > 0) {
            // Iterate over a copy so that `once` listeners can remove themselves safely.
            for (const listener of [...listeners]) {
                const result = listener.fn(current);
                if (listener.once) {
                    const index = listeners.indexOf(listener);
                    if (index >= 0) {
                        listeners.splice(index, 1);
                    }
                }
                if (result) {
                    if (result.message !== undefined) {
                        current = { ...current, message: result.message };
                    }
                    if (result.preventDefault) {
                        preventDefault = true;
                    }
                }
            }
        }
        return { message: current, preventDefault };
    }
    /**
     * Notifies the host of a message.
     * The caller waits until the notification completes.
     */
    async notify(msg) {
        await this.maybeEmitTelemetry(msg);
        // Run any registered listeners. A listener may update the message text
        // and/or prevent the default processing (e.g. stack-activity messages are
        // routed to the activity printer and not written to a stream).
        const { message, preventDefault } = this.applyMessageListeners(msg);
        if (preventDefault) {
            return;
        }
        if (!(0, api_private_1.isMessageRelevantForLevel)(message, this.logLevel)) {
            return;
        }
        if (this.corkedCounter > 0) {
            this.corkedLoggingBuffer.push(message);
            return;
        }
        const output = this.formatMessage(message);
        const stream = this.selectStream(message);
        stream?.write(output);
    }
    async maybeEmitTelemetry(msg) {
        try {
            const telemetryEvent = eventFromMessage(msg);
            if (telemetryEvent) {
                await this.telemetry?.emit(telemetryEvent);
            }
        }
        catch (e) {
            await this.defaults.trace(`Emit Telemetry Failed ${e.message}`);
        }
    }
    /**
     * Route stack-activity messages to the activity printer (progress bar or
     * event list) rather than writing them to a stream.
     *
     * Implemented as listeners that handle the message via the printer and
     * prevent the default processing, so the rest of the pipeline does not also
     * emit them. The printer is created lazily on the first stack-activity message.
     */
    routeStackActivityToPrinter() {
        const route = (msg) => {
            if (!this.activityPrinter) {
                this.activityPrinter = this.makeActivityPrinter();
            }
            this.activityPrinter.notify(msg);
            return { preventDefault: true }; // handled by the printer; don't also write to a stream
        };
        this.on(api_private_1.IO.CDK_TOOLKIT_I5501, route);
        this.on(api_private_1.IO.CDK_TOOLKIT_I5502, route);
        this.on(api_private_1.IO.CDK_TOOLKIT_I5503, route);
    }
    /**
     * Detect special messages encode information about whether or not
     * they require approval
     */
    skipApprovalStep(msg) {
        const approvalToolkitCodes = ['CDK_TOOLKIT_I5060'];
        if (!(msg.code && approvalToolkitCodes.includes(msg.code))) {
            return false;
        }
        switch (this.requireDeployApproval) {
            // Never require approval
            case cloud_assembly_schema_1.RequireApproval.NEVER:
                return true;
            // Always require approval
            case cloud_assembly_schema_1.RequireApproval.ANYCHANGE:
                return false;
            // Require approval if changes include broadening permissions
            case cloud_assembly_schema_1.RequireApproval.BROADENING:
                return ['none', 'non-broadening'].includes(msg.data?.permissionChangeType);
        }
    }
    /**
     * Determines the output stream, based on message and configuration.
     */
    selectStream(msg) {
        if (isNoticesMessage(msg)) {
            return targetStreamObject(this.noticesDestination);
        }
        return this.selectStreamFromLevel(msg.level);
    }
    /**
     * Determines the output stream, based on message level and configuration.
     */
    selectStreamFromLevel(level) {
        // The stream selection policy for the CLI is the following:
        //
        //   (1) Messages of level `result` always go to `stdout`
        //   (2) Messages of level `error` always go to `stderr`.
        //   (3a) All remaining messages go to `stderr`.
        //   (3b) If we are in CI mode, all remaining messages go to `stdout`.
        //
        switch (level) {
            case 'error':
                return process.stderr;
            case 'result':
                return process.stdout;
            default:
                return this.isCI ? process.stdout : process.stderr;
        }
    }
    /**
     * Notifies the host of a message that requires a response.
     *
     * If the host does not return a response the suggested
     * default response from the input message will be used.
     */
    async requestResponse(msg) {
        // If the request cannot be prompted for by the CliIoHost, we just accept the default
        if (!isPromptableRequest(msg)) {
            await this.notify(msg);
            return msg.defaultResponse;
        }
        const response = await this.withCorkedLogging(async () => {
            // prepare prompt data
            // @todo this format is not defined anywhere, probably should be
            const data = msg.data ?? {};
            const motivation = data.motivation ?? 'User input is needed';
            const concurrency = data.concurrency ?? 0;
            const responseDescription = data.responseDescription;
            // Special approval prompt
            // Determine if the message needs approval. If it does, continue (it is a basic confirmation prompt)
            // If it does not, return success (true). We only check messages with codes that we are aware
            // are requires approval codes.
            if (this.skipApprovalStep(msg)) {
                return true;
            }
            // In --yes mode, respond for the user if we can
            if (this.autoRespond) {
                // respond with yes to all confirmations
                if (isConfirmationPrompt(msg)) {
                    await this.notify({
                        ...msg,
                        message: `${chalk.cyan(msg.message)} (auto-confirmed)`,
                    });
                    return true;
                }
                // respond with the default for all other messages
                if (msg.defaultResponse) {
                    await this.notify({
                        ...msg,
                        message: `${chalk.cyan(msg.message)} (auto-responded with default: ${util.format(msg.defaultResponse)})`,
                    });
                    return msg.defaultResponse;
                }
            }
            // only talk to user if STDIN is a terminal (otherwise, fail)
            if (!this.isTTY) {
                throw new toolkit_lib_1.ToolkitError('TtyNotAttached', `${motivation}, but terminal (TTY) is not attached so we are unable to get a confirmation from the user`);
            }
            // only talk to user if concurrency is 1 (otherwise, fail)
            if (concurrency > 1) {
                throw new toolkit_lib_1.ToolkitError('ConcurrencyConflict', `${motivation}, but concurrency is greater than 1 so we are unable to get a confirmation from the user`);
            }
            // Basic confirmation prompt
            // We treat all requests with a boolean response as confirmation prompts
            if (isConfirmationPrompt(msg)) {
                const confirmed = await promptly.confirm(`${chalk.cyan(msg.message)} (y/n)`);
                if (!confirmed) {
                    throw new toolkit_lib_1.ToolkitError('AbortedByUser', 'Aborted by user');
                }
                return confirmed;
            }
            // Asking for a specific value
            const prompt = extractPromptInfo(msg);
            const desc = responseDescription ?? prompt.default;
            const answer = await promptly.prompt(`${chalk.cyan(msg.message)}${desc ? ` (${desc})` : ''}`, {
                default: prompt.default,
                trim: true,
            });
            return prompt.convertAnswer(answer);
        });
        // We need to cast this because it is impossible to narrow the generic type
        // isPromptableRequest ensures that the response type is one we can prompt for
        // the remaining code ensure we are indeed returning the correct type
        return response;
    }
    /**
     * Formats a message for console output with optional color support
     */
    formatMessage(msg) {
        // apply provided style or a default style if we're in TTY mode
        let message_text = this.isTTY
            ? styleMap[msg.level](msg.message)
            : msg.message;
        // prepend timestamp if IoMessageLevel is DEBUG or TRACE. Postpend a newline.
        return ((msg.level === 'debug' || msg.level === 'trace')
            ? `[${this.formatTime(msg.time)}] ${message_text}`
            : message_text) + '\n';
    }
    /**
     * Formats date to HH:MM:SS
     */
    formatTime(d) {
        const pad = (n) => n.toString().padStart(2, '0');
        return `${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`;
    }
    /**
     * Get an instance of the ActivityPrinter
     */
    makeActivityPrinter() {
        const props = {
            stream: this.selectStreamFromLevel('info'),
        };
        switch (this.stackProgress) {
            case deploy_1.StackActivityProgress.EVENTS:
                return new api_private_1.HistoryActivityPrinter(props);
            case deploy_1.StackActivityProgress.BAR:
                return new api_private_1.CurrentActivityPrinter(props);
        }
    }
}
exports.CliIoHost = CliIoHost;
/**
 * This IoHost implementation considers a request promptable, if:
 * - it's a yes/no confirmation
 * - asking for a string or number value
 */
function isPromptableRequest(msg) {
    return isConfirmationPrompt(msg)
        || typeof msg.defaultResponse === 'string'
        || typeof msg.defaultResponse === 'number';
}
/**
 * Check if the request is a confirmation prompt
 * We treat all requests with a boolean response as confirmation prompts
 */
function isConfirmationPrompt(msg) {
    return typeof msg.defaultResponse === 'boolean';
}
/**
 * Helper to extract information for promptly from the request
 */
function extractPromptInfo(msg) {
    const isNumber = (typeof msg.defaultResponse === 'number');
    const defaultResponse = util.format(msg.defaultResponse);
    return {
        default: defaultResponse,
        defaultDesc: 'defaultDescription' in msg && msg.defaultDescription ? util.format(msg.defaultDescription) : defaultResponse,
        convertAnswer: isNumber ? (v) => Number(v) : (v) => String(v),
    };
}
const styleMap = {
    error: chalk.red,
    warn: chalk.yellow,
    result: chalk.reset,
    info: chalk.reset,
    debug: chalk.gray,
    trace: chalk.gray,
};
function targetStreamObject(x) {
    switch (x) {
        case 'stderr':
            return process.stderr;
        case 'stdout':
            return process.stdout;
        case 'drop':
            return undefined;
    }
}
function isNoticesMessage(msg) {
    return api_private_1.IO.CDK_TOOLKIT_I0100.is(msg) || api_private_1.IO.CDK_TOOLKIT_W0101.is(msg) || api_private_1.IO.CDK_TOOLKIT_E0101.is(msg) || api_private_1.IO.CDK_TOOLKIT_I0101.is(msg);
}
function eventFromMessage(msg) {
    if (messages_1.CLI_PRIVATE_IO.CDK_CLI_I1001.is(msg)) {
        return eventResult('SYNTH', msg);
    }
    if (messages_1.CLI_PRIVATE_IO.CDK_CLI_I2001.is(msg)) {
        return eventResult('INVOKE', msg);
    }
    if (messages_1.CLI_PRIVATE_IO.CDK_CLI_I3001.is(msg)) {
        return eventResult('DEPLOY', msg);
    }
    if (messages_1.CLI_PRIVATE_IO.CDK_CLI_I3003.is(msg)) {
        return eventResult('ASSET', msg);
    }
    // Hotswap lives in the cdk-toolkit so it cannot be a CDK_CLI error code.
    // Instead we reuse the existing Hotswap span.
    if (api_private_1.IO.CDK_TOOLKIT_I5410.is(msg)) {
        // Create a telemetry-compatible result
        return hotswapToEventResult(msg.data);
    }
    return undefined;
    function eventResult(eventType, m) {
        return {
            eventType,
            duration: m.data.duration,
            error: m.data.error,
            counters: m.data.counters,
        };
    }
}
function hotswapToEventResult(result) {
    const nonHotswappableResources = {};
    for (const { subject } of result.nonHotswappableChanges) {
        if ('resourceType' in subject) {
            const keys = 'rejectedProperties' in subject && subject.rejectedProperties
                ? subject.rejectedProperties.map(p => `hotswapFallback:${subject.resourceType}#${p}`)
                : [`hotswapFallback:${subject.resourceType}`];
            for (const key of keys) {
                nonHotswappableResources[key] = (nonHotswappableResources[key] ?? 0) + 1;
            }
        }
    }
    return {
        eventType: 'HOTSWAP',
        duration: result.duration,
        ...(result.error ? {
            error: {
                name: (0, error_1.cdkCliErrorName)(result.error),
            },
        } : {}),
        counters: {
            hotswapped: result.hotswapped ? 1 : 0,
            hotswapFallback: result.hotswapFallback ? 1 : 0,
            hotswappableChanges: result.hotswappableChanges.length,
            nonHotswappableChanges: result.nonHotswappableChanges.length,
            ...nonHotswappableResources,
        },
    };
}
